x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2023 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.wildfly.security.certificate.management.x500.cert.acme;
+
+import io.smallrye.mutiny.Uni;
+import org.wildfly.security.certificate.management.x500.cert.X509CertificateChainAndSigningKey;
+
+import java.net.URL;
+import java.security.PrivateKey;
+import java.security.cert.CRLReason;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import java.util.Map;
+
+public abstract class AsyncAcmeClientSpi {
+
+    private final AcmeClientSpi delegate;
+
+    public AsyncAcmeClientSpi() {
+        AsyncAcmeClientSpi impl = this;
+
+        this.delegate = new AcmeClientSpi() {
+            @Override
+            public AcmeChallenge proveIdentifierControl(AcmeAccount account, List<AcmeChallenge> challenges) throws AcmeException {
+                return impl.proveIdentifierControl(account, challenges);
+            }
+
+            @Override
+            public void cleanupAfterChallenge(AcmeAccount account, AcmeChallenge challenge) throws AcmeException {
+                impl.cleanupAfterChallenge(account, challenge);
+            }
+        };
+    }
+
+    protected abstract AcmeChallenge proveIdentifierControl(AcmeAccount account, List<AcmeChallenge> challenges) throws AcmeException;
+    protected abstract void cleanupAfterChallenge(AcmeAccount account, AcmeChallenge challenge) throws AcmeException;
+
+    public Uni<Map<AcmeResource, URL>> getResourceUrls(ElytronRequestContext context, AcmeAccount account, boolean staging) throws AcmeException {
+        return context.runBlocking(() -> {
```

Comment:
I'm kind of concerned by this approach. It should be the opposite. Async first, blocking on top. Not the opposite. This is not only pretty bad in term of performance, it can lead to many issues (propagation, ordering...).

Reply:
@cescoffier Thank you very much for the review and for the idea! I thought about this, but I actually have some concerns now regarding doing the opposite, as it would mean changing the current stable API implementation logic to utilize the mutiny API.  It seems to me that doing so could be worth it, only if some of the logic could be invoked in concurrent threads to achieve a performance benefit.

- There are actually not many places that can be made more performant as they are usually a simple requests to a server. This API will also be used by EAP and should work for any ACME certificate authority. One candidate method that could be made more performant is a method `obtainCertificateChain` that can send a single request to obtain a certificate for multiple domains/subdomains. If we were to optimise this and send requests in parallel threads, there is a risk that the ACME CA implementation could have network limits based on which they could drop such requests. Dropping the connections and thus failing to obtaining the certificate could break existing clients/users. They already document rate limits that would cause them to drop the requests[1], but these could also change in the future and/or be different for other ACME CAs. 
- There could also be new bugs introduced with this change that would also break existing users and would add maintainance costs in the future.

The bottleneck of all these API methods is the network connection, as each method makes a request to a server, so wrapping the synchronous API does not seem to worsen the performance in a significant way. Also these methods are not invoked very often as they relate to rotation of the certificates.

What are your thoughts on this?

[1] https://letsencrypt.org/docs/rate-limits/ , see for example the limit for this: 
> The “new-nonce”, “new-account”, “new-order”, and “revoke-cert” endpoints on the API have an Overall Requests limit of 20 per second. The “/directory” endpoint and the “/acme” directory & subdirectories have an Overall Requests limit of 40 requests per second.

Reply:
The problem is that the current approach will limit the concurrency of the quarkus applications to 20 (default number of workers), and the application will suffer from multiple thread jumps. Even if there is no concurrent call to Acme, it will be a severe bottleneck and will underperform (like really, really slow).

Each executeBlocking leads to 2 thread jumps, and the second one moves the task to the end of the event loop queue.  

So, maybe we need to have 2 APIs: one sync and one async, which would use non-blocking I/O under the hood and avoid the executeBlocking jumps.

Reply:
@cescoffier Okay thank you, that makes sense. I will look into this

Reply:
Hello @cescoffier , I have started to rewrite the async API to use asynchronous calls with Uni. I created a draft PR [here](https://github.com/wildfly-security-incubator/certificate-management/pull/3). 

The important class is the [AyncAcmeClientSpi](https://github.com/wildfly-security-incubator/certificate-management/blob/c4a7586f4637206d74966b96bad02aaf87528b16/x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java) class. The methods now use [HttpClient.sendAsync](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.html#sendAsync(java.net.http.HttpRequest,java.net.http.HttpResponse.BodyHandler,java.net.http.HttpResponse.PushPromiseHandler)) and use/chain Uni calls when needed. 

Not all methods are rewritten yet, but I wanted to make sure that this approach will be suitable for quarkus ?

Thank you!